### PR TITLE
monitoring: monitoring for DB and ES

### PR DESCRIPTION
**Pull request by rerowep**
* Adds functionality to monitoring values from DB and ES.

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- https://tree.taiga.io/project/rero21-reroils/task/1342

## How to test?

after setup:
- website: localhost:5000/nonitoring/info
- cli: pipenv run invenio monitoring info

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: rero_ils/modules/monitoring.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            if index:
+                msg += '\t{index:>16}: {count_es:>10}'.format(
+                    index=index,
+                    count_es=info.get('es', '')
+                )
+            result += msg + '\n'
+        return result
+
+    @classmethod
+    def _get_all_endpoints(cls):
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: useless

File: rero_ils/modules/monitoring.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        """Get all endpoints from config."""
+        return current_app.config.get('RECORDS_REST_ENDPOINTS')
+
+    @classmethod
+    def _get_search_index(cls, endpoint):
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: useless

File: rero_ils/modules/monitoring.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        """Get search index for doc type from config."""
+        return endpoint.get('search_index', '')
+
+    @classmethod
+    def get_db_count(cls, doc_type, with_deleted=False):
+        """Get the DB count."""
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: Can you describe the parameters and the return value in you docstring?

File: rero_ils/modules/monitoring.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        return query
+
+    @classmethod
+    def get_es_count(cls, index):
+        """Get the DB count."""
+        if not index:
+            return('Not in ES')
+        try:
+            result = RecordsSearch(index=index).query().count()
+        except:
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: You need to check the exception type.

File: rero_ils/modules/monitoring.py (hunk 1 above)
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: deleted

File: rero_ils/modules/monitoring.py (hunk 2 above)
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: deleted

File: rero_ils/modules/monitoring.py (hunk 4 above)
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: ok

File: tests/ui/test_monitoring.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+from utils import flush_index
+
+from rero_ils.modules.documents.api import Document, DocumentsSearch
+from rero_ils.modules.monitoring import Monitoring
+
+
+def test_monitoring(app, document_sion_items_data):
+    """Test monitoring."""
+    mon = Monitoring()
+    assert mon.get_es_count('xxx') == 'No >>xxx<< in ES'
```
Comment (jma) [inline tests/ui/test_monitoring.py]: you can do the same with get_db_count

File: rero_ils/modules/monitoring.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+api_blueprint = Blueprint(
+    'api_monitoring',
+    __name__,
+    url_prefix='/monitoring'
+)
+
+
+@api_blueprint.route('/es_db_counts')
+def info():
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: ok

File: rero_ils/modules/monitoring.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    """Display ES and DB count informations."""
+    return jsonify({'data': Monitoring.info()})
+
+
+@api_blueprint.route('/check_es_db_counts')
+def check():
+    """Check ES and DB counts."""
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: ok

File: rero_ils/modules/monitoring.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        else:
+            data['DB'].append(pid)
+    for pid in mon.get('ES'):
+        if api_url:
+            data['ES'].append(api_url + pid)
+        else:
+            data['ES'].append(pid)
+    for pid in mon.get('ES double'):
+        if api_url:
+            url = api_url + '?q=pid:' + pid
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: ok

File: rero_ils/modules/persons/api.py [hunk 9]
```diff
@@ -86,17 +88,6 @@ def get_record_by_mef_pid(cls, pid):
         rec.reindex()
         return rec
 
-        rec = cls.get_record_by_pid(pid)
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/persons/api.py]: This is useless code after `return rec` and nice to have it cleaned out.

[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/persons/api.py]: ok

File: tests/ui/test_monitoring.py (hunk 5 above)
Comment (rerowep) [reply tests/ui/test_monitoring.py]: ok

File: rero_ils/modules/monitoring.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                msg += '\t{index:>16}: {count_es:>10}'.format(
+                    index=index,
+                    count_es=info.get('es', '')
+                )
+            result += msg + '\n'
+        return result
+
+    @classmethod
+    def get_db_count(cls, doc_type, with_deleted=False):
+        """Get the DB count.
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: I tried, I hope it is better now.

File: rero_ils/modules/monitoring.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+            'invenio_records_rest.{doc_type}_list'.format(doc_type=doc_type),
+            _external=True
+        )
+    except:
+        api_url = None
+    mon = Monitoring.missing(doc_type)
+    data = {'DB': [], 'ES': [], 'ES double': {}}
+    for pid in mon.get('DB'):
+        if api_url:
+            data['DB'].append(api_url + '?q=pid:' + pid)
```
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/monitoring.py]: ok

File: rero_ils/modules/monitoring.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+            result += msg + '\n'
+        return result
+
+    @classmethod
+    def get_db_count(cls, doc_type, with_deleted=False):
+        """Get DB item count.
+
+        :param doc_type: document type to get count from DB.
+        :param with_deleted: count also deleted items in DB.
+        :retrun: item count for document type in DB.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: retrun -> return

File: rero_ils/modules/monitoring.py [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+api_blueprint = Blueprint(
+    'api_monitoring',
+    __name__,
+    url_prefix='/monitoring'
+)
+
+
+@api_blueprint.route('/es_db_counts')
+def es_db_counts():
+    """Displays informations for ES and DB counts."""
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: which kind of information?

File: rero_ils/modules/monitoring.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+            return jsonify({'status': 'error: Forbidden'}), 403
+        return func(*args, **kwargs)
+
+    return decorated_view
+
+
+@api_blueprint.route('/missing_pids/<doc_type>')
+@check_authentication
+def missing_pids(doc_type):
+    """Displays missing pids for doc type.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: missing where and why?

File: rero_ils/modules/monitoring.py [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+            count = data['ES double'].setdefault(url, 0)
+            data['ES double'][url] = count + 1
+        else:
+            count = data['ES double'].setdefault(pid, 0)
+            data['ES double'][pid] = count + 1
+    return jsonify({'data': data})
+
+
+class Monitoring(object):
+    """Monitoring class."""
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: can you put more details in the comment. The main idea here is to check the consistency between the database and the search index. We need to check that all documents presents in the database are also present in the search index and vice versa.

File: rero_ils/modules/monitoring.py [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+                msg += '\t{index:>16}: {count_es:>10}'.format(
+                    index=index,
+                    count_es=info.get('es', '')
+                )
+            result += msg + '\n'
+        return result
+
+    @classmethod
+    def get_db_count(cls, doc_type, with_deleted=False):
+        """Get DB item count.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: DB -> database

File: rero_ils/modules/monitoring.py [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+        query = PersistentIdentifier.query.filter(
+            PersistentIdentifier.pid_type == doc_type
+        )
+        if not with_deleted:
+            query = query.filter_by(status=PIDStatus.REGISTERED)
+        return query.count()
+
+    @classmethod
+    def get_es_count(cls, index):
+        """Get ES item count.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: ES -> elasticsearch index or search index

File: rero_ils/modules/monitoring.py [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+            return('Not in ES')
+        try:
+            result = RecordsSearch(index=index).query().count()
+        except NotFoundError:
+            result = 'No >>{index}<< in ES'.format(index=index)
+        return result
+
+    @classmethod
+    def info(cls, with_deleted=False):
+        """Get count details for all records rest endpoints in json format.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: count of what?

File: rero_ils/modules/monitoring.py [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+
+        :param with_deleted: count also deleted items in DB.
+        :retruns: dictionair with DB ES and DB-ES count informations.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: spell

File: rero_ils/modules/monitoring.py [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+            if db_es not in [0, '']:
+                checks[info] = db_es
+        return checks
+
+    @classmethod
+    def missing(cls, doc_type):
+        """Get missing pids in DB and ES and find double pids in ES.
+
+        :param doc_type: doc type to get missing pids.
+        :retruns: dictionair with all missing pids.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: spell

File: rero_ils/modules/monitoring.py [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+        checks = {}
+        for info, data in cls.info().items():
+            db_es = data.get('db-es', '')
+            if db_es not in [0, '']:
+                checks[info] = db_es
+        return checks
+
+    @classmethod
+    def missing(cls, doc_type):
+        """Get missing pids in DB and ES and find double pids in ES.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: double -> duplicate

File: rero_ils/modules/monitoring.py (hunk 20 above)
Comment (jma) [inline rero_ils/modules/monitoring.py]: spell

File: rero_ils/modules/monitoring.py [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+        missing_in_es = set(pids_db).difference(set(pids_es))
+        return {
+            'DB': list(missing_in_db),
+            'ES': list(missing_in_es),
+            'ES double': [x for x in pids_es if pids_es.count(x) > 1]
+        }
+
+    @classmethod
+    def print_missing(cls, doc_type):
+        """Print missing pids for doc type.
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: doc type -> the given document type

File: rero_ils/modules/monitoring.py [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+def monitoring():
+    """Monitoring commands."""
+
+
+@monitoring.command('es_db_counts')
+@click.option('-m', '--missing', 'missing', is_flag=True, default=False,
+              help='display missing pids')
+@with_appcontext
+def es_db_counts_cli(missing):
+    """Print ES and DB counts."""
```
Comment (jma) [inline rero_ils/modules/monitoring.py]: need more explanation

File: rero_ils/permissions.py [hunk 24]
```diff
@@ -30,6 +30,7 @@
 
 request_item_permission = DynamicPermission(RoleNeed('patron'))
 librarian_permission = DynamicPermission(RoleNeed('librarian'))
+admin_permission = DynamicPermission(RoleNeed('admin'))
```
Comment (jma) [inline rero_ils/permissions.py]: You can replace DynamicPermission by Permission as the former is legacy.

**Pull request merged**